flax/nnx/nn/recurrent.py

```diff
@@ -0,0 +1,920 @@
+        rngs: rnglib.Rngs,
+    ):
+        self.in_features = in_features
+        self.hidden_features = hidden_features
+        self.gate_fn = gate_fn
+        self.activation_fn = activation_fn
+        self.kernel_init = kernel_init
+        self.recurrent_kernel_init = recurrent_kernel_init
+        self.bias_init = bias_init
+        self.dtype = dtype
+        self.param_dtype = param_dtype
+        self.carry_init = carry_init
+        self.rngs = rngs
+
+        # input and recurrent layers are summed so only one needs a bias.
+        dense_i = partial(
+            Linear,
+            in_features=in_features,
+            out_features=hidden_features,
+            use_bias=False,
+            kernel_init=self.kernel_init,
+            dtype=self.dtype,
+            param_dtype=self.param_dtype,
+            rngs=rngs,
+        )
+
+        dense_h = partial(
+            Linear,
+            in_features=hidden_features,
+            out_features=hidden_features,
+            use_bias=True,
+            kernel_init=self.recurrent_kernel_init,
+            bias_init=self.bias_init,
+            dtype=self.dtype,
+            param_dtype=self.param_dtype,
+            rngs=rngs,
+        )
+
+        self.ii = dense_i()
+        self.if_ = dense_i()
+        self.ig = dense_i()
+        self.io = dense_i()
+        self.hi = dense_h()
+        self.hf = dense_h()
+        self.hg = dense_h()
+        self.ho = dense_h()
+
+    def __call__(self, carry: tuple[Array, Array], inputs: Array) -> tuple[tuple[Array, Array], Array]:
+        r"""A long short-term memory (LSTM) cell.
+
+        Args:
+          carry: the hidden state of the LSTM cell,
+            initialized using ``LSTMCell.initialize_carry``.
+          inputs: an ndarray with the input for the current time step.
+            All dimensions except the final are considered batch dimensions.
+
+        Returns:
+          A tuple with the new carry and the output.
+        """
+        c, h = carry
+        i = self.gate_fn(self.ii(inputs) + self.hi(h))
+        f = self.gate_fn(self.if_(inputs) + self.hf(h))
+        g = self.activation_fn(self.ig(inputs) + self.hg(h))
+        o = self.gate_fn(self.io(inputs) + self.ho(h))
+        new_c = f * c + i * g
+        new_h = o * self.activation_fn(new_c)
+        return (new_c, new_h), new_h
+
+    def initialize_carry(
+        self, rngs: rnglib.Rngs | None, input_shape: tuple[int, ...]
+    ) -> tuple[Array, Array]:
+        """Initialize the RNN cell carry.
+
+        Args:
+          rng: random number generator passed to the init_fn.
+          input_shape: a tuple providing the shape of the input to the cell.
+        Returns:
+          An initialized carry for the given RNN cell.
+        """
+        batch_dims = input_shape[:-1]
+        if rngs is None:
+            rngs = self.rngs
+        mem_shape = batch_dims + (self.hidden_features,)
+        c = self.carry_init(rngs.params(), mem_shape, self.param_dtype)
```

Comment:
Maybe we could use the `carry` stream as these are not technically trainable params?
```suggestion
        c = self.carry_init(rngs.carry(), mem_shape, self.param_dtype)
```

Reply:
Same for all other cells.